pkg/policycache/cache.go

```diff
@@ -42,5 +43,55 @@ func (c *cache) GetPolicies(pkey PolicyType, kind, nspace string) []kyvernov1.Po
 		result = append(result, c.store.get(pkey, kind, nspace)...)
 		result = append(result, c.store.get(pkey, "*", nspace)...)
 	}
+
+	if pkey == ValidateAudit { // also get policies with ValidateEnforce
+		result = append(result, c.store.get(ValidateEnforce, kind, "")...)
+		result = append(result, c.store.get(ValidateEnforce, "*", "")...)
+	}
+
+	if pkey == ValidateAudit || pkey == ValidateEnforce {
+		result = filterPolicies(pkey, result, nspace, kind)
+	}
+
 	return result
 }
+
+// filter for cluster policies on validationFailureAction is overriden
+func filterPolicies(pkey PolicyType, result []kyvernov1.PolicyInterface, nspace, kind string) []kyvernov1.PolicyInterface {
+	var policies []kyvernov1.PolicyInterface
+	for _, policy := range result {
+		validationFailureAction := policy.GetSpec().ValidationFailureAction
+		keepPolicy := true
+		overrides := policy.GetSpec().ValidationFailureActionOverrides
+
+		if pkey == ValidateAudit {
+			if validationFailureAction == kyvernov1.Enforce && (len(overrides) == 0 || nspace == "") {
+				keepPolicy = false
+			} else {
+				for _, action := range overrides {
+					if action.Action == kyvernov1.Enforce && kyvernoutils.ContainsNamepace(action.Namespaces, nspace) {
+						keepPolicy = false
+						break
+					}
+				}
+			}
+
+		} else if pkey == ValidateEnforce {
+			if validationFailureAction == kyvernov1.Audit && nspace == "" {
+				keepPolicy = false
+			} else {
```

Comment:
This a lot of if-else for one function, needs some refactoring. Like "pkey" is being used in if and else-if conditions but there is no final else block that suggests that it can have other values. Can we use switch-case here?

Reply:
refactored code